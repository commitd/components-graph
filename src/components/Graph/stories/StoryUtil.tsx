import React from 'react'
import { Graph } from '../'
import {
  addRandomEdge,
  addRandomNode,
  GraphModel,
  cytoscapeRenderer,
  GraphRendererOptions,
} from '../../../graph'

export const exampleModel = addRandomEdge(
  addRandomNode(GraphModel.createEmpty(), 20),
  15
)

export const Template: React.FC<{
  model: GraphModel
  onModelChange?: (
    model: GraphModel | ((model2: GraphModel) => GraphModel)
  ) => void
  height?: GraphRendererOptions['height']
}> = ({ model, onModelChange, height }) => {
  return (
    <Graph
      model={model}
      onModelChange={onModelChange}
      renderer={cytoscapeRenderer}
      options={{ height: height ?? 600 }}
    />
  )
}
