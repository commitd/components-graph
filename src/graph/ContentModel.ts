import { v4 } from 'uuid'
import {
  ModelAttributeSet,
  ModelEdge,
  ModelGraphData,
  ModelNode,
} from './types'

export class ContentModel {
  public static fromRaw(data: ModelGraphData): ContentModel {
    let model = Object.values(data.nodes).reduce(
      (acc, node) => acc.addNode(node),
      ContentModel.createEmpty()
    )
    model = Object.values(data.edges).reduce(
      (acc, edge) => acc.addEdge(edge),
      model
    )
    return model
  }

  public static createEmpty(): ContentModel {
    return new ContentModel({}, {})
  }

  constructor(
    readonly nodes: Record<string, ModelNode>,
    readonly edges: Record<string, ModelEdge>
  ) {
    this.nodes = nodes
    this.edges = edges
  }

  containsNode(id: string): boolean {
    return this.getNode(id) != null
  }

  containsEdge(id: string): boolean {
    return this.getEdge(id) != null
  }

  getNode(id: string): ModelNode | undefined {
    return this.nodes[`${id}`]
  }

  private getExistingNode(id: string): ModelNode {
    const node = this.getNode(id)
    if (node == null) {
      throw new Error(`Node [${id}] does not exist`)
    }
    return node
  }

  getEdgesLinkedToNode(nodeId: string): ModelEdge[] {
    return Object.values(this.edges).filter(
      (e) => e.target === nodeId || e.source === nodeId
    )
  }

  addNode(node: Partial<ModelNode>): ContentModel {
    if (node.id != null && this.containsNode(node.id)) {
      throw new Error(`Cannot add node already in graph (${node.id})`)
    }
    const newNode: ModelNode = {
      id: node.id ?? v4(),
      attributes: node.attributes ?? {},
      ...node,
    }
    const nodes = { ...this.nodes, [newNode.id]: newNode }
    return new ContentModel(nodes, this.edges)
  }

  editNode(node: ModelNode): ContentModel {
    this.getExistingNode(node.id)
    return new ContentModel(
      { ...this.nodes, ...{ [node.id]: node } },
      this.edges
    )
  }

  addNodeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    const node = this.getExistingNode(id)
    const newAttributes = {
      ...node.attributes,
      ...{ [attributeName]: attributeValue },
    }
    const changedNode = {
      ...this.getExistingNode(id),
      ...{ attributes: newAttributes },
    }
    return this.editNode(changedNode)
  }

  /**
   * Remove node and associated edges
   * @param id id of the node
   */
  removeNode(id: string): ContentModel {
    const node = this.getNode(id)
    if (node == null) {
      return this
    }
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    let content: ContentModel = this
    this.getEdgesLinkedToNode(id).forEach((e) => {
      content = content.removeEdge(e.id)
    })
    return content.removeOrphanNode(id)
  }

  private removeOrphanNode(id: string): ContentModel {
    const withoutNode = { ...this.nodes }
    delete withoutNode[`${id}`]
    return new ContentModel(withoutNode, this.edges)
  }

  editNodeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    if (this.getExistingNode(id).attributes[`${attributeName}`] == null) {
      throw new Error(`Node [${id}] does not have attribute ${attributeName}`)
    }
    return this.addNodeAttribute(id, attributeName, attributeValue)
  }

  removeNodeAttribute(id: string, attributeName: string): ContentModel {
    const node = this.getExistingNode(id)
    const remainingAttributes = { ...node.attributes }
    delete remainingAttributes[`${attributeName}`]
    const n = { ...node, ...{ attributes: remainingAttributes } }
    return this.editNode(n)
  }

  addEdge(
    edge: Omit<ModelEdge, 'id' | 'attributes'> & {
      id?: string
      attributes?: ModelAttributeSet
    }
  ): ContentModel {
    // check source and target exist
    this.getExistingNode(edge.source)
    this.getExistingNode(edge.target)
    if (edge.id != null && this.containsEdge(edge.id)) {
      throw new Error(`Cannot add edge already in graph (${edge.id})`)
    }
    const newEdge: ModelEdge = {
      id: edge.id ?? v4(),
      attributes: edge.attributes ?? {},
      ...edge,
    }
    const edges = { ...this.edges, [newEdge.id]: newEdge }
    return new ContentModel(this.nodes, edges)
  }

  getEdge(id: string): ModelEdge | undefined {
    return this.edges[`${id}`]
  }

  private getExistingEdge(id: string): ModelEdge {
    const edge = this.getEdge(id)
    if (edge == null) {
      throw new Error(`Edge [${id}] does not exist`)
    }
    return edge
  }

  editEdge(edge: ModelEdge): ContentModel {
    // ensure edge exists
    this.getExistingEdge(edge.id)
    return new ContentModel(this.nodes, {
      ...this.edges,
      ...{ [edge.id]: edge },
    })
  }

  addEdgeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    const edge = this.getExistingEdge(id)
    const newAttributes = {
      ...edge.attributes,
      ...{ [attributeName]: attributeValue },
    }
    const changedEdge = {
      ...this.getExistingEdge(id),
      ...{ attributes: newAttributes },
    }
    return this.editEdge(changedEdge)
  }

  editEdgeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    if (this.getExistingEdge(id).attributes[`${attributeName}`] == null) {
      throw new Error(`Edge [${id}] does not have attribute ${attributeName}`)
    }
    return this.addEdgeAttribute(id, attributeName, attributeValue)
  }

  removeEdgeAttribute(id: string, attributeName: string): ContentModel {
    const edge = this.getExistingEdge(id)
    const remainingAttributes = { ...edge.attributes }
    delete remainingAttributes[`${attributeName}`]
    const e = { ...edge, ...{ attributes: remainingAttributes } }
    return this.editEdge(e)
  }

  removeEdge(id: string): ContentModel {
    const edge = this.getEdge(id)
    if (edge == null) {
      return this
    }
    const withoutEdge = { ...this.edges }
    delete withoutEdge[`${id}`]
    return new ContentModel(this.nodes, withoutEdge)
  }
}
