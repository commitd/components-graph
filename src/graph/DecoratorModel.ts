import { colors, Theme } from '@committed/components'
import {
  DecoratedEdge,
  DecoratedNode,
  DecorationFunction,
  EdgeDecoration,
  EdgeDecorationFunction,
  EdgeDecorator,
  ItemDecoration,
  ModelEdge,
  ModelNode,
  NodeDecoration,
  NodeDecorationFunction,
  NodeDecorator,
} from './types'

export class DecoratorModel {
  static readonly DEFAULT_NODE_DECORATION: NodeDecoration = {
    shape: 'ellipse',
    color: '#FFc526',
    label: '',
    size: 25,
    opacity: 1,
    strokeColor: '#3E3E3E',
    strokeSize: 2,
  }

  static readonly DEFAULT_EDGE_DECORATION: EdgeDecoration = {
    color: '#FF0000',
    size: 2,
    label: '',
    sourceArrow: false,
    targetArrow: false,
    opacity: 1,
    style: 'solid',
  }

  static readonly idAsLabelNode: NodeDecorator = (item) => {
    return { label: item.id }
  }

  static readonly idAsLabelNodeEdge: EdgeDecorator = (item) => {
    return { label: item.id }
  }

  static createDefault(
    options: {
      nodeDecorators?: NodeDecorator[]
      nodeDefaults?: Partial<NodeDecoration> | NodeDecorationFunction
      edgeDecorators?: EdgeDecorator[]
      edgeDefaults?: Partial<EdgeDecoration> | EdgeDecorationFunction
    } = {}
  ): DecoratorModel {
    return new DecoratorModel(
      options.nodeDecorators ?? [],
      options.nodeDefaults ?? {},
      options.edgeDecorators ?? [],
      options.edgeDefaults ?? {}
    )
  }

  constructor(
    private readonly nodeDecorators: NodeDecorator[],
    private readonly nodeDefaults:
      | Partial<NodeDecoration>
      | NodeDecorationFunction,
    private readonly edgeDecorators: EdgeDecorator[],
    private readonly edgeDefaults:
      | Partial<EdgeDecoration>
      | EdgeDecorationFunction
  ) {
    this.nodeDecorators = nodeDecorators
    this.nodeDefaults = nodeDefaults
    this.edgeDecorators = edgeDecorators
    this.edgeDefaults = edgeDefaults
  }

  /**
   * Get the decoration that is applied to all nodes, before anything is overridden by decorators
   *
   * Use this to apply blanket styling to nodes to minimise the performance impact of per-node styling
   */
  getDefaultNodeDecorator(): DecorationFunction<
    NodeDecoration,
    NodeDecoration
  > {
    return (theme: Theme) => ({
      ...DecoratorModel.DEFAULT_NODE_DECORATION,
      ...{
        strokeColor: theme.palette.text.secondary,
        color: colors.committedYellow[500],
      },
      ...DecoratorModel.wrapDecorator(this.nodeDefaults)(theme),
    })
  }

  /**
   * Get the decoration that is applied to all edges, before anything is overridden by decorators
   *
   * Use this to apply blanket styling to edges to minimise the performance impact of per-edge styling
   */
  getDefaultEdgeDecorator(): DecorationFunction<
    EdgeDecoration,
    EdgeDecoration
  > {
    return (theme: Theme) => ({
      ...DecoratorModel.DEFAULT_EDGE_DECORATION,
      ...{
        color: theme.palette.text.secondary,
      },
      ...DecoratorModel.wrapDecorator(this.edgeDefaults)(theme),
    })
  }

  /**
   * Adds a function to the given nodes that provides the decoration overrides for the node
   * @param modelNodes to be decorated
   */
  getDecoratedNodes(modelNodes: ModelNode[]): DecoratedNode[] {
    return Object.values(modelNodes).map((node) => {
      return {
        getDecorationOverrides: (theme: Theme) => ({
          ...this.getNodeDecorationOverrides(node, theme),
        }),
        ...node,
      }
    })
  }

  /**
   * Adds a function to the given edges that provides the decoration overrides for the edge
   * @param modelEdges to be decorated
   */
  getDecoratedEdges(modelEdges: ModelEdge[]): DecoratedEdge[] {
    return Object.values(modelEdges).map((edge) => {
      return {
        getDecorationOverrides: (theme: Theme) => ({
          ...this.getEdgeDecorationOverrides(edge, theme),
        }),
        ...edge,
      }
    })
  }

  private getNodeDecorationOverrides(
    node: ModelNode,
    theme: Theme
  ): Partial<NodeDecoration> {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { id, attributes, ...nodeStyle } = node
    const decor = Object.assign(
      {},
      ...this.nodeDecorators.map((d) => d(node, theme))
    ) as Partial<NodeDecoration>
    return {
      ...decor,
      ...nodeStyle,
    }
  }

  private getEdgeDecorationOverrides(
    edge: ModelEdge,
    theme: Theme
  ): Partial<EdgeDecoration> {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { id, attributes, source, target, ...edgeStyle } = edge
    const decor = Object.assign(
      {},
      ...this.edgeDecorators.map((d) => d(edge, theme))
    ) as Partial<EdgeDecoration>
    return {
      ...decor,
      ...edgeStyle,
    }
  }

  private static isFunction<T extends ItemDecoration>(
    decorator: Partial<T> | DecorationFunction<T>
  ): decorator is DecorationFunction<T> {
    return typeof decorator === 'function'
  }

  private static wrapDecorator<T extends ItemDecoration>(
    decorator: Partial<T> | DecorationFunction<T>
  ): DecorationFunction<T> {
    if (DecoratorModel.isFunction(decorator)) {
      return decorator
    } else {
      return () => decorator
    }
  }
}
