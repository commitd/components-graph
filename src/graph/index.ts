export * from './ContentModel'
export * from './GraphModel'
export * from './DecoratorModel'
export * from './LayoutModel'
export * from './SelectionModel'
export * from './data'
export * from './types'
export * from './renderer'
