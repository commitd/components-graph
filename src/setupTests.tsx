import { ThemeProvider } from '@committed/components'
import { StylesProvider } from '@material-ui/styles'
import '@testing-library/jest-dom/extend-expect'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GenerateId } from 'jss'
import React from 'react'

const generateClassName: GenerateId = (rule, styleSheet) => {
  const prefix =
    styleSheet === undefined || styleSheet.options.classNamePrefix === undefined
      ? ''
      : styleSheet.options.classNamePrefix
  return `${prefix}-${rule.key}`
}

const LightTheme: React.FC = ({ children }) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="light">{children}</ThemeProvider>
  </StylesProvider>
)

const DarkTheme: React.FC = ({ children }) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="dark">{children}</ThemeProvider>
  </StylesProvider>
)

export const renderPlain = render

export const renderLight = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: LightTheme, ...options })

export const renderDark = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: DarkTheme, ...options })

// re-export everything
export * from '@testing-library/react'
export { userEvent }
