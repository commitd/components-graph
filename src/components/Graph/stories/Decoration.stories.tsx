import { Theme } from '@committed/components'
import { Meta } from '@storybook/react'
import React, { useState } from 'react'
import { Graph } from '../'
import { ContentModel, DecoratorModel, GraphModel } from '../../../graph'
import {
  addRandomEdge,
  addRandomNodeColors,
  addRandomNodeShapes,
} from '../../../graph/data/Generator'
import { ModelNode, NodeDecoration } from '../../../graph/types'
import { exampleGraphData } from './exampleData'
import { Template } from './StoryUtil'

export default {
  title: 'Examples/Decoration',
  component: Graph,
  decorators: [
    (story) => <div style={{ height: '50vh', padding: '16px' }}>{story()}</div>,
  ],
  parameters: {
    layout: 'fullscreen',
  },
} as Meta

export const NodeShapes: React.FC = () => {
  const [model, setModel] = useState(
    addRandomEdge(addRandomNodeShapes(GraphModel.createEmpty(), 20), 15)
  )
  return <Template model={model} onModelChange={setModel} />
}

export const NodeColors: React.FC = () => {
  const [model, setModel] = useState(
    addRandomEdge(addRandomNodeColors(GraphModel.createEmpty(), 20), 15)
  )
  return <Template model={model} onModelChange={setModel} />
}

export const TypedDecoration: React.FC = () => {
  const [model, setModel] = useState(
    new GraphModel(ContentModel.fromRaw(exampleGraphData), {
      decoratorModel: DecoratorModel.createDefault({
        nodeDecorators: [
          (node: ModelNode, theme: Theme): Partial<NodeDecoration> => {
            const type = node.attributes['type']
            if (type === 'person') return { color: theme.palette.info.main }
            if (type === 'place') return { color: theme.palette.success.main }
            return { color: '#00FF00' }
          },
        ],
      }),
    })
  )

  return <Template model={model} onModelChange={setModel} />
}

export const CustomIcons: React.FC = () => {
  const [model, setModel] = useState(
    GraphModel.createWithContent(
      new ContentModel(
        {
          chris: {
            id: 'chris',
            attributes: {},
            label: 'Chris Flatley',
            icon: 'https://committed.software/static/chris-e5738250e4525dd9c6783ea3802c814f.jpg',
            size: 100,
          },
          stuart: {
            id: 'stuart',
            attributes: {},
            label: 'Stuart Hendren',
            icon: 'https://committed.software/static/stuart-faed1d5684aa9dd6d7cc20fff9e34e8e.png',
            size: 100,
          },
          jon: {
            id: 'jon',
            attributes: {},
            label: 'Jon Elliot',
            icon: 'https://committed.software/static/jon-809f2f941386ebb56949c0b3f5dae221.png',
            size: 100,
          },
          steve: {
            id: 'steve',
            attributes: {},
            label: 'Steven Taylor',
            icon: 'https://committed.software/static/steve-70f3e007ab54a82ef7c7ab8a7989ebc3.jpg',
            size: 100,
          },
          matt: {
            id: 'matt',
            attributes: {},
            label: 'Matt Copas',
            icon: 'https://committed.software/static/matt-2db182f3e2a621017e7ef4cffda2ee3f.jpg',
            size: 100,
          },
          kristian: {
            id: 'kristian',
            attributes: {},
            label: 'Kristian Aspinall',
            icon: 'https://committed.software/static/kristian-4508fcd63e07690867e4c39b51bdd685.jpg',
            size: 100,
          },
        },
        {
          e1: {
            id: 'e1',
            attributes: {},
            source: 'chris',
            target: 'stuart',
          },
          e2: {
            id: 'e2',
            attributes: {},
            source: 'stuart',
            target: 'jon',
          },
          e3: {
            id: 'e3',
            attributes: {},
            source: 'kristian',
            target: 'steve',
          },
          e4: {
            id: 'e4',
            attributes: {},
            source: 'matt',
            target: 'kristian',
          },
          e5: {
            id: 'e5',
            attributes: {},
            source: 'jon',
            target: 'matt',
          },
          e6: {
            id: 'e6',
            attributes: {},
            source: 'jon',
            target: 'steve',
          },
          e7: {
            id: 'e7',
            attributes: {},
            source: 'chris',
            target: 'steve',
          },
        }
      )
    )
  )
  return <Template model={model} onModelChange={setModel} />
}
