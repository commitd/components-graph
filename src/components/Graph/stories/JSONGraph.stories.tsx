import { Meta } from '@storybook/react'
import React, { useState } from 'react'
import { Graph } from '..'
import { ContentModel, GraphModel } from '../../../graph'
import { largeGraph, smallGraph } from '../../../graph/data/jsonGraphExamples'
import { Template } from './StoryUtil'

export default {
  title: 'Examples/JSONGraph',
  component: Graph,
  decorators: [
    (story) => <div style={{ height: '100h', padding: '16px' }}>{story()}</div>,
  ],
  parameters: {
    layout: 'fullscreen',
  },
} as Meta

export const SimpleJsonGraph: React.FC = () => {
  const [model, setModel] = useState(
    GraphModel.createWithContent(ContentModel.fromJsonGraph(smallGraph))
  )
  return <Template model={model} onModelChange={setModel} />
}

export const RichJsonGraph: React.FC = () => {
  const [model, setModel] = useState(
    GraphModel.createWithContent(ContentModel.fromJsonGraph(largeGraph))
  )
  return <Template model={model} onModelChange={setModel} />
}
