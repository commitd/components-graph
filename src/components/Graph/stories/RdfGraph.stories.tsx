import { Alert, Flex } from '@committed/components'
import { Meta, Story } from '@storybook/react'
import React, { useState, useEffect } from 'react'
import { Graph } from '..'
import {
  ContentModel,
  cytoscapeRenderer,
  GraphBuilder,
  GraphModel,
  ModelNode,
} from '../../../graph'
import { DECORATION_IRI, LiteralOption } from '../../../graph/fromRdfGraph'
import { GraphToolbar } from '../../GraphToolbar'
import { NodeViewer } from '../../NodeViewer'
import { Template } from './StoryUtil'

export default {
  title: 'Examples/RdfGraph',
  component: Graph,
  decorators: [
    (story) => <div style={{ height: '100h', padding: '16px' }}>{story()}</div>,
  ],
  parameters: {
    layout: 'fullscreen',
  },
} as Meta

export const SmallGraph: React.FC = () => {
  const [model, setModel] = useState(
    GraphModel.createWithContent(
      GraphBuilder.fromRdfGraph(
        `
      :John a :Man ;
        :name "John" ;
        :hasSpouse :Mary .
      :Mary a :Woman ;
        :name "Mary" ;
        :hasSpouse :John .
      :John_jr a :Man ;
        :name "John Jr." ;
        :hasParent :John, :Mary .
      :Time_Span a owl:Class .
      :event a :Activity ;
      :has_time_span [
        a :Time_Span ;
        :at_some_time_within_date "2018-01-12"^^xsd:date
      ] .
      :u129u-klejkajo-2309124u-sajfl a :Person ;
        :name "John Doe" .
  `,
        {
          additionalPrefixes: {
            '': '',
            owl: 'http://www.w3.org/2002/07/owl#',
            xsd: 'http://www.w3.org/2001/XMLSchema#',
          },
        }
      )
    )
  )
  return <Template model={model} onModelChange={setModel} />
}

export const SmallGraphWithTypes: React.FC = () => {
  const [model, setModel] = useState(
    GraphModel.createWithContent(
      GraphBuilder.fromRdfGraph(
        `
    :John a :Man ;
      :name "John" ;
      :hasSpouse :Mary .
    :Mary a :Woman ;
      :name "Mary" ;
      :hasSpouse :John .
    :John_jr a :Man ;
      :name "John Jr." ;
      :hasParent :John, :Mary .
    :Time_Span a owl:Class .
    :event a :Activity ;
    :has_time_span [
      a :Time_Span ;
      :at_some_time_within_date "2018-01-12"^^xsd:date
    ] .
    :u129u-klejkajo-2309124u-sajfl a :Person ;
      :name "John Doe" .
`,
        {
          type: undefined,
          additionalPrefixes: {
            '': '',
            owl: 'http://www.w3.org/2002/07/owl#',
            xsd: 'http://www.w3.org/2001/XMLSchema#',
          },
        }
      )
    )
  )
  return <Template model={model} onModelChange={setModel} />
}

export const DecoratedGraph: React.FC = () => {
  const [model, setModel] = useState(
    GraphModel.createWithContent(
      GraphBuilder.fromRdfGraph(`
    @prefix cd: <${DECORATION_IRI}> .
  
    <A> a <Test> ;
        cd:label "Test" ;
        cd:size 10 ;
        cd:color '#FFBB00' ;
        cd:opacity 0.5 ;
        cd:shape 'diamond' ;
        cd:strokeColor '#000000' ;
        cd:strokeSize 1 .
  `)
    )
  )
  return <Template model={model} onModelChange={setModel} />
}

const StoryTemplate: Story<{ rdf: string }> = ({ rdf }) => {
  const [model, setModel] = useState(
    GraphModel.createWithContent(ContentModel.createEmpty())
  )
  const [alert, setAlert] = useState<string | undefined>()

  useEffect(() => {
    try {
      setModel(
        GraphModel.createWithContent(
          GraphBuilder.fromRdfGraph(rdf, {
            literals: LiteralOption.VALUE_ONLY,
          })
        )
      )
      setAlert(undefined)
    } catch (error) {
      setModel(GraphModel.createWithContent(ContentModel.createEmpty()))
      setAlert(error.message)
    }
  }, [setModel, setAlert, rdf])

  const [node, setNode] = useState<ModelNode | undefined>(undefined)

  return (
    <>
      {alert && (
        <Alert color="error" title={'RDF Error'}>
          {alert}
        </Alert>
      )}
      <Flex>
        <GraphToolbar
          flexDirection="column"
          model={model}
          onModelChange={setModel}
          layouts={cytoscapeRenderer.layouts}
        />
        <Graph
          model={model}
          onModelChange={setModel}
          renderer={cytoscapeRenderer}
          options={{ height: 600 }}
          onViewNode={setNode}
        />
        <NodeViewer node={node} onClose={() => setNode(undefined)} />
      </Flex>
    </>
  )
}

export const RdfVisualizer = StoryTemplate.bind({})
RdfVisualizer.args = {
  rdf: `
  @prefix cd: <${DECORATION_IRI}> .

  <A> a <Test> ;
      cd:label "Test" ;
      cd:size 10 ;
      cd:color '#FFBB00' ;
      cd:opacity 0.5 ;
      cd:shape 'diamond' ;
      cd:strokeColor '#000000' ;
      cd:strokeSize 1 .
`,
}
