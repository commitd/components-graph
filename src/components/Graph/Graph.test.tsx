import React from 'react'
import { renderLight, renderDark } from '../../setupTests'
import { CustomRenderer } from './stories/CustomRenderer.stories'
import { ContentModel, GraphModel } from '../../graph'
import { exampleGraphData } from './stories/exampleData'

it('renders light', () => {
  const { asFragment } = renderLight(
    <CustomRenderer
      initialModel={new GraphModel(ContentModel.fromRaw(exampleGraphData))}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})
it('renders dark', () => {
  const { asFragment } = renderDark(
    <CustomRenderer
      initialModel={new GraphModel(ContentModel.fromRaw(exampleGraphData))}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})
