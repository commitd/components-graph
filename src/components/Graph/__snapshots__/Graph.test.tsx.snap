// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dark 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Node
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Edge
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Nodes
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "n1",
    "attributes": {
      "type": "person"
    }
  },
  {
    "id": "n2",
    "attributes": {
      "type": "person"
    }
  },
  {
    "id": "n3",
    "attributes": {
      "type": "person"
    }
  },
  {
    "id": "n4",
    "attributes": {
      "type": "place"
    }
  },
  {
    "id": "n5",
    "attributes": {
      "type": "place"
    }
  },
  {
    "id": "n6",
    "attributes": {
      "type": "place"
    }
  }
]
      </pre>
    </div>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Edges
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "e1",
    "attributes": {},
    "source": "n1",
    "target": "n2"
  },
  {
    "id": "e2",
    "attributes": {},
    "source": "n1",
    "target": "n3"
  },
  {
    "id": "e3",
    "attributes": {},
    "source": "n3",
    "target": "n4"
  },
  {
    "id": "e5",
    "attributes": {},
    "source": "n5",
    "target": "n5"
  }
]
      </pre>
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`renders light 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Node
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Edge
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Nodes
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "n1",
    "attributes": {
      "type": "person"
    }
  },
  {
    "id": "n2",
    "attributes": {
      "type": "person"
    }
  },
  {
    "id": "n3",
    "attributes": {
      "type": "person"
    }
  },
  {
    "id": "n4",
    "attributes": {
      "type": "place"
    }
  },
  {
    "id": "n5",
    "attributes": {
      "type": "place"
    }
  },
  {
    "id": "n6",
    "attributes": {
      "type": "place"
    }
  }
]
      </pre>
    </div>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Edges
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "e1",
    "attributes": {},
    "source": "n1",
    "target": "n2"
  },
  {
    "id": "e2",
    "attributes": {},
    "source": "n1",
    "target": "n3"
  },
  {
    "id": "e3",
    "attributes": {},
    "source": "n3",
    "target": "n4"
  },
  {
    "id": "e5",
    "attributes": {},
    "source": "n5",
    "target": "n5"
  }
]
      </pre>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
